master/internal/resourcemanagers/priority.go

```diff
@@ -0,0 +1,331 @@
+				priorityToScheduledTaskMap, toRelease)
+
+			if taskPlaced {
+				localAgentsState = updatedLocalAgentState
+				for preemptedTask := range preemptedTasks {
+					log.Infof("preempting task %s for task %s",
+						preemptedTask.Address().Local(), prioritizedAllocation.Name)
+					toRelease[preemptedTask] = true
+				}
+			}
+		}
+	}
+
+	// Convert to a slice which is the expected output format.
+	toReleaseSlice := make([]*actor.Ref, 0, len(toRelease))
+	for k := range toRelease {
+		toReleaseSlice = append(toReleaseSlice, k)
+	}
+
+	return toAllocate, toReleaseSlice
+}
+
+// trySchedulingTaskViaPreemption checks whether preempting lower priority tasks
+// would allow this task to be scheduled.
+func trySchedulingTaskViaPreemption(
+	taskList *taskList,
+	allocationRequest *AllocateRequest,
+	allocationPriority int,
+	fittingMethod SoftConstraint,
+	agents map[*actor.Ref]*agentState,
+	priorityToScheduledTaskMap map[int][]*AllocateRequest,
+	tasksAlreadyPreempted map[*actor.Ref]bool,
+) (bool, map[*actor.Ref]*agentState, map[*actor.Ref]bool) {
+	localAgentsState := deepCopyAgents(agents)
+	preemptedTasks := make(map[*actor.Ref]bool)
+	log.Debugf("trying to schedule task %s by preempting other tasks", allocationRequest.Name)
+
+	for priority := model.MaxUserSchedulingPriority; priority > allocationPriority; priority-- {
+		if _, ok := priorityToScheduledTaskMap[priority]; !ok {
+			continue
+		}
+
+		preemptionCandidates := priorityToScheduledTaskMap[priority]
+		for _, preemptionCandidate := range preemptionCandidates {
+			if preemptionCandidate.NonPreemptible || preemptionCandidate.SlotsNeeded == 0 {
+				continue
+			}
+
+			if _, ok := tasksAlreadyPreempted[preemptionCandidate.TaskActor]; ok {
+				continue
+			}
+
+			resourcesAllocated := taskList.GetAllocations(preemptionCandidate.TaskActor)
+			removeTaskFromAgents(localAgentsState, resourcesAllocated)
+			preemptedTasks[preemptionCandidate.TaskActor] = true
+
+			if fits := findFits(allocationRequest, localAgentsState, fittingMethod); len(fits) > 0 {
+				addTaskToAgents(fits)
+				return true, localAgentsState, preemptedTasks
+			}
+		}
+	}
+
+	return false, localAgentsState, preemptedTasks
+}
+
+// trySchedulingPendingTasksInPriority tries to schedule all the tasks in the
+// current priority. Note tasks are scheduled based on the order in which they
+// are listed.
+func trySchedulingPendingTasksInPriority(
+	allocationRequests []*AllocateRequest,
+	agents map[*actor.Ref]*agentState,
+	fittingMethod SoftConstraint,
+) ([]*AllocateRequest, []*AllocateRequest) {
+	successfulAllocations := make([]*AllocateRequest, 0)
+	unSuccessfulAllocations := make([]*AllocateRequest, 0)
+
+	for _, allocationRequest := range allocationRequests {
+		fits := findFits(allocationRequest, agents, fittingMethod)
+		if len(fits) == 0 {
+			unSuccessfulAllocations = append(unSuccessfulAllocations, allocationRequest)
+			continue
```

Comment:
blocking: if we are unable to allocate - is the correct behavior to break and stop allocating, even if there is room? (letting preemption, if true, clear space as needed)

Reply:
We want to try and schedule all the tasks we can in this priority before we start using preemption to clear space, because tasks that are scheduled without preemption can be started right away, while tasks that need preemption don't get started until the preemption completes.

Reply:
Ah, so then we aren't enforcing timestamps like we enforce priorities? e.g. t1 task must be scheduled before t2 task

Reply:
Yep exactly, the timestamps are just the order in which we consider tasks, but not a hard ordering constraint.